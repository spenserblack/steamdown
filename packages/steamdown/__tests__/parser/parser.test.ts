import { describe, expect, test } from "@jest/globals";
import Parser, { Heading } from "../../src/parser";

describe("Parser", () => {
  describe(".next()", () => {
    describe.each([
      ["# Heading 1", Heading],
      ["## Heading 2", Heading],
      ["### Heading 3", Heading],
      ["#### Heading 4", Heading],
      ["##### Heading 5", Heading],
      ["###### Heading 6", Heading],
      ["# Heading 1\n", Heading],
      ["## Heading 2\n", Heading],
      ["### Heading 3\n", Heading],
      ["#### Heading 4\n", Heading],
      ["##### Heading 5\n", Heading],
      ["###### Heading 6\n", Heading],
      ["# *Italic heading*", Heading],
      ["# _Italic heading_", Heading],
      ["# *Partially* italic heading", Heading],
      ["# *Italic heading* with *italic text*", Heading],
      ["# Heading with *italic tail*", Heading],
    ])("when given %p", (text, tokenType) => {
      test("returns a token", () => {
        const parser = new Parser(text);
        const token = parser.next();
        expect(token).toBeInstanceOf(tokenType);
      });

      test("inline tokens", () => {
        const parser = new Parser(text);
        const token = parser.next();
        expect(token.tokens).toMatchSnapshot();
      });

      test("rendered token", () => {
        const parser = new Parser(text);
        const token = parser.next();
        expect(token.render()).toMatchSnapshot();
      });
    });
  });
});
