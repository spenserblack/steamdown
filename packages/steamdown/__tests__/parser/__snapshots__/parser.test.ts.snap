// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser .next() when given "# *Italic heading* with *italic text*" inline tokens 1`] = `
[
  Italics {
    "content": "Italic heading",
    "raw": "*Italic heading*",
    "tag": "i",
    "tokens": [
      InlineText {
        "content": "Italic heading",
        "raw": "Italic heading",
      },
    ],
  },
  InlineText {
    "content": " with ",
    "raw": " with ",
  },
  Italics {
    "content": "italic text",
    "raw": "*italic text*",
    "tag": "i",
    "tokens": [
      InlineText {
        "content": "italic text",
        "raw": "italic text",
      },
    ],
  },
]
`;

exports[`Parser .next() when given "# *Italic heading* with *italic text*" rendered token 1`] = `"[h1][i]Italic heading[/i] with [i]italic text[/i][/h1]"`;

exports[`Parser .next() when given "# *Italic heading*" inline tokens 1`] = `
[
  Italics {
    "content": "Italic heading",
    "raw": "*Italic heading*",
    "tag": "i",
    "tokens": [
      InlineText {
        "content": "Italic heading",
        "raw": "Italic heading",
      },
    ],
  },
]
`;

exports[`Parser .next() when given "# *Italic heading*" rendered token 1`] = `"[h1][i]Italic heading[/i][/h1]"`;

exports[`Parser .next() when given "# *Partially* italic heading" inline tokens 1`] = `
[
  Italics {
    "content": "Partially",
    "raw": "*Partially*",
    "tag": "i",
    "tokens": [
      InlineText {
        "content": "Partially",
        "raw": "Partially",
      },
    ],
  },
  InlineText {
    "content": " italic heading",
    "raw": " italic heading",
  },
]
`;

exports[`Parser .next() when given "# *Partially* italic heading" rendered token 1`] = `"[h1][i]Partially[/i] italic heading[/h1]"`;

exports[`Parser .next() when given "# _Italic heading_" inline tokens 1`] = `
[
  Italics {
    "content": "Italic heading",
    "raw": "_Italic heading_",
    "tag": "i",
    "tokens": [
      InlineText {
        "content": "Italic heading",
        "raw": "Italic heading",
      },
    ],
  },
]
`;

exports[`Parser .next() when given "# _Italic heading_" rendered token 1`] = `"[h1][i]Italic heading[/i][/h1]"`;

exports[`Parser .next() when given "# Heading 1
" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 1",
    "raw": "Heading 1",
  },
]
`;

exports[`Parser .next() when given "# Heading 1
" rendered token 1`] = `"[h1]Heading 1[/h1]"`;

exports[`Parser .next() when given "# Heading 1" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 1",
    "raw": "Heading 1",
  },
]
`;

exports[`Parser .next() when given "# Heading 1" rendered token 1`] = `"[h1]Heading 1[/h1]"`;

exports[`Parser .next() when given "# Heading with *italic tail*" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading with ",
    "raw": "Heading with ",
  },
  Italics {
    "content": "italic tail",
    "raw": "*italic tail*",
    "tag": "i",
    "tokens": [
      InlineText {
        "content": "italic tail",
        "raw": "italic tail",
      },
    ],
  },
]
`;

exports[`Parser .next() when given "# Heading with *italic tail*" rendered token 1`] = `"[h1]Heading with [i]italic tail[/i][/h1]"`;

exports[`Parser .next() when given "## Heading 2
" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 2",
    "raw": "Heading 2",
  },
]
`;

exports[`Parser .next() when given "## Heading 2
" rendered token 1`] = `"[h2]Heading 2[/h2]"`;

exports[`Parser .next() when given "## Heading 2" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 2",
    "raw": "Heading 2",
  },
]
`;

exports[`Parser .next() when given "## Heading 2" rendered token 1`] = `"[h2]Heading 2[/h2]"`;

exports[`Parser .next() when given "### Heading 3
" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 3",
    "raw": "Heading 3",
  },
]
`;

exports[`Parser .next() when given "### Heading 3
" rendered token 1`] = `"[h3]Heading 3[/h3]"`;

exports[`Parser .next() when given "### Heading 3" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 3",
    "raw": "Heading 3",
  },
]
`;

exports[`Parser .next() when given "### Heading 3" rendered token 1`] = `"[h3]Heading 3[/h3]"`;

exports[`Parser .next() when given "#### Heading 4
" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 4",
    "raw": "Heading 4",
  },
]
`;

exports[`Parser .next() when given "#### Heading 4
" rendered token 1`] = `"[h4]Heading 4[/h4]"`;

exports[`Parser .next() when given "#### Heading 4" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 4",
    "raw": "Heading 4",
  },
]
`;

exports[`Parser .next() when given "#### Heading 4" rendered token 1`] = `"[h4]Heading 4[/h4]"`;

exports[`Parser .next() when given "##### Heading 5
" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 5",
    "raw": "Heading 5",
  },
]
`;

exports[`Parser .next() when given "##### Heading 5
" rendered token 1`] = `"[h5]Heading 5[/h5]"`;

exports[`Parser .next() when given "##### Heading 5" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 5",
    "raw": "Heading 5",
  },
]
`;

exports[`Parser .next() when given "##### Heading 5" rendered token 1`] = `"[h5]Heading 5[/h5]"`;

exports[`Parser .next() when given "###### Heading 6
" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 6",
    "raw": "Heading 6",
  },
]
`;

exports[`Parser .next() when given "###### Heading 6
" rendered token 1`] = `"[h6]Heading 6[/h6]"`;

exports[`Parser .next() when given "###### Heading 6" inline tokens 1`] = `
[
  InlineText {
    "content": "Heading 6",
    "raw": "Heading 6",
  },
]
`;

exports[`Parser .next() when given "###### Heading 6" rendered token 1`] = `"[h6]Heading 6[/h6]"`;
